### DISCUSSION TITLE
Support for HSMs and/or non-extractable private keys?

### DISCUSSION THREAD

**Discussion by coder5876** [panva/jose#66] [Q&A]
First off, thank you so much for building/maintaining this great library! 😃👍

Question: In the signing functions for JWS exposed in this library it is required to pass in the private key into the function in order to sign a specific JWT. Are there any plans to also support signing with private keys that cannot be extracted, such as those created on HSMs or secure enclaves? Or is that deliberately out of scope for this library?

**Comment by panva** [comment]
It is out of scope, at least until node supports engines natively. It’s built on top of all node’s `crypto` module can offer. Nothing more.

**Comment by coder5876** [comment]
Thanks for the answer @panva, that makes sense!
